Controller forced reconfiguration: pare down

joe-redpanda opened the pull request:
Node-wise recovery today does not support force moving partitions which are already moving.

As a result, the existing implementation of controller forced recovery would often trigger partition moves from the brokers to be decommissioned before the force_moves from node-wise recovery were attempted.

The longterm resolution is to make node-wise recovery override existing moves.

For now, though, CFR is substantially more reliable if the operators execute node-wise recovery and then decommission.

This PR pares back controller forced recovery to only execute the raft0 update and wait for a leader, alongside the relevant test updates per the new expected operator workflow.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Bug Fixes
* makes CFR cleanup manual to avoid nodewise recovery race
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda committed:
cluster/cfr: test changes per scaled back cfr:

controller_forced_reconfiguration will, for now, only recover the
controller leader.

The burden of executing nodewise recovery and decomission will
then fall on the caller.

This commit updates the fixture tests per this expectation such that
they execute nodewise recovery and decommission after controller forced
reconfiguration.

joe-redpanda committed:
rpk.py: chore typing improvement

joe-redpanda committed:
cfr: scale back cfr ducktape test

The burden of calling nodewise recovery and decomission now falls on the
operator. This commit updates controller_forced_reconfiguration_test.py
to match those expectations. It now executes forced recoonfiguration,
executes nodewise recovery, decommissions the dead brokers, and then
checks cluster health.

joe-redpanda committed:
cluster/cfr_m: remove nodewise recovery and decom

Removes nodewise recovery and node decommission from the steps that
controller forced recovery executes.

These, for now, will need to be executed by the operator.

joe-redpanda committed:
breakglass: documentation updates

Updates the breakglass service documentation per the changes to
controller forced recovery.

Removes indications of nodewise recovery and decomissioning as
controller_forced_recovery will no longer perform these.

joe-redpanda committed:
admin.py: chore typing improvements

joe-redpanda commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -328,10 +334,150 @@ class CFRFixture
+                return model::broker_shard{.node_id = node_id, .shard = 0};
+            })};
+
+        // request the movwe
```
I'll fix this if I need to push again.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2c9cfa0f-c177-4dc9-953c-03336ec4d91d-->
<!--TEST RESULTS-->
<!--a046cbbd-8567-4aff-9afc-617c88d1b728-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75443">build#75443</a></summary>


| test_class                           | test_method                          | test_arguments                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:-------------------------------------|:-------------------------------------|:--------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests                 | test_task_pausing                    | {"shuffle_leadership": false}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75443#019a3c8c-9fb6-41ac-ada6-5ef880f3a235 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_task_pausing                          |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync           | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/75443#019a3c84-1628-4864-9627-55864f23c9db | FLAKY         | 15/21    | upstream reliability is '96.17590822179733'. current run reliability is '71.42857142857143'. drift is 24.74734 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkingReplicationTests        | test_replication_timestamps_match    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75443#019a3c84-1624-43e5-902c-1610ba015c14 | FLAKY         | 20/21    | upstream reliability is '95.13888888888889'. current run reliability is '95.23809523809523'. drift is -0.09921 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| ShadowLinkingReplicationTests        | test_replication_timestamps_match    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75443#019a3c8c-9fad-43ed-8ba5-519972e385b3 | FLAKY         | 20/21    | upstream reliability is '95.14563106796116'. current run reliability is '95.23809523809523'. drift is -0.09246 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| DataMigrationsApiTest                | test_creating_and_listing_migrations | null                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/75443#019a3c8c-9fad-43ed-8ba5-519972e385b3 | FLAKY         | 19/21    | upstream reliability is '96.15720524017468'. current run reliability is '90.47619047619048'. drift is 5.68101 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations      |
| topic_properties_syncer_test         | topic_properties_sync                |                                                                                       | unit        | https://buildkite.com/redpanda/redpanda/builds/75443#019a3bce-ebd0-4cab-92d1-bf37791042f2 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75935">build#75935</a></summary>


| test_class                          | test_method                | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                              |
|:------------------------------------|:---------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionE2ETest | test_crash_all             | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/75935#019a6f2b-2d03-449a-b997-f6fb5415befa | FLAKY         | 13/21    | upstream reliability is '91.03641456582633'. current run reliability is '61.904761904761905'. drift is 29.13165 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all          |
| WriteCachingFailureInjectionTest    | test_unavoidable_data_loss | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/75935#019a6f30-53c6-4793-a19a-de6f8ca21cf3 | FLAKY         | 20/21    | upstream reliability is '94.90616621983914'. current run reliability is '95.23809523809523'. drift is -0.33193 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<!--/body-->

bashtanov commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -328,10 +334,150 @@ class CFRFixture
+                            .get();
+        ASSERT_TRUE(maybe_pwlm.has_value())
+          << "get partitions_with_lost_majority failed with error: "
+          << maybe_pwlm.assume_error().message();
```
nit: IIRC a result may contain neither a value nor an error, so use `.error()`

bashtanov replied:
this comment hasn't been addressed

joe-redpanda replied:
not sure how I missed this, apologies

bashtanov commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -328,10 +334,150 @@ class CFRFixture
-        // and that [ceil(N/2), N + ceil(N/2)] are alive
+        // given a cluster size of N, check that [0, floor(N/2)] are
+        // actually gone (killed to lose quorum) and that [ceil(N/2), N +
+        // ceil(N/2)] are alive
```
nit: accidental reflow after indent+outdent? it's more readable when formulae aren't wrapped

joe-redpanda replied:
yeah i'll fix this

bashtanov commented on tests/rptest/tests/controller_forced_reconfiguration_test.py:
```diff
@@ -15,6 +15,7 @@
 from ducktape.tests.test import TestContext
 from ducktape.utils.util import wait_until
 
+from rptest.clients.rpk import RpkTool
```
nit: typo in the file name in the commit message

bashtanov replied:
this comment hasn't been addressed

bashtanov commented on tests/rptest/tests/controller_forced_reconfiguration_test.py:
```diff
@@ -317,12 +331,25 @@ def controller_available():
+
+        self._check_topic_recovered(topic, cluster_size, killed_node_ids)
+
         KgoVerifierProducer.oneshot(  # type: ignore
```
https://github.com/redpanda-data/redpanda/pull/28261 will add type hints to it

bashtanov commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -488,6 +642,10 @@ TEST_F(CFRFixture3, Smoke3) {
+    execute_nodewise_recovery();
+
+    decommission_dead_nodes();
+
```
nit: double line breaks all over look excessive to me

bashtanov reviewed (commented):
LGTM, a few nits

bashtanov commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -328,9 +334,149 @@ class CFRFixture
-        // gone (killed to lose quorum)
+        // given a cluster size of N, check that all nodes
+        // [0, floor(N/2)] are actually gone (killed to lose quorum)
         // and that [ceil(N/2), N + ceil(N/2)] are alive
```
nit: technically only correct for odd-sized cluster

joe-redpanda replied:
Good catch. I think this justifies throwing in an even cluster number as well.

joe-redpanda committed:
controller_forced_reconfiguration_test: even tests

The code paths between even cluster sizes and odd cluster sizes are
marginally different. Adds even cluster size test cases.

Moves the smoke test orchestration code into the test fixture to reduce
copy paste.

bashtanov commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -403,6 +574,33 @@ TEST(CalculateRaftGroup, TwoSuvivors) {
+
+    ASSERT_EQ(result_config.size(), expected.size());
+    for (const auto& result_node : result_config) {
+        ASSERT_TRUE(std::ranges::find(expected, result_node) != expected.end());
```
nit: `ASSERT_TRUE(std::ranges::contains(expected, result_node);`

joe-redpanda replied:
Merging for now, i'll take this as a todo in the future for CFR gets an update to add back nodewise recovery and node decom

vbotbuildovich commented:
/backport v25.3.x

vbotbuildovich commented:
Branch name "v25.3.x" not found.

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/19271782337).
